File: apps/robot/drivers/src/rc_pulse_motor_driver.cpp
// Copyright 2026 Mumtahin Farabi
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included in
// all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
// THE SOFTWARE.


#include <unistd.h>

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <fstream>
#include <memory>
#include <stdexcept>
#include <string>
#include <thread>

#include <rclcpp/rclcpp.hpp>
#include <robot_platform_msgs/msg/drive.hpp>
#include <robot_platform_msgs/msg/feedback.hpp>

using robot_platform_msgs::msg::Drive;
using robot_platform_msgs::msg::Feedback;

constexpr int PWM_EXPORT_RETRIES = 500;
constexpr auto PWM_EXPORT_RETRY_DELAY = std::chrono::milliseconds(10);
constexpr double FEEDBACK_RATE_HZ = 50.0;
constexpr double COMMAND_TIMEOUT_SECONDS = 0.5;
constexpr int64_t PULSE_PERIOD_NS = 20'000'000;
constexpr int64_t PULSE_NEUTRAL_NS = 1'500'000;
constexpr int64_t PULSE_HALF_RANGE_NS = 500'000;

void write_sysfs(const std::string & path, const std::string & value)
{
  std::ofstream file(path);
  file << value;
  file.flush();
  if (!file.good()) {
    throw std::runtime_error("failed writing " + value + " to " + path);
  }
}

class PulseChannel
{
public:
  PulseChannel(int64_t chip, int64_t channel, bool reversed)
  : reversed(reversed),
    channel_path(
      "/sys/class/pwm/pwmchip" + std::to_string(chip) + "/pwm" + std::to_string(channel))
  {
    const std::string duty_path = channel_path + "/duty_cycle";
    if (access(channel_path.c_str(), F_OK) != 0) {
      write_sysfs(
        "/sys/class/pwm/pwmchip" + std::to_string(chip) + "/export", std::to_string(channel));
    }
    for (int attempt = 0; attempt < PWM_EXPORT_RETRIES; ++attempt) {
      if (access(duty_path.c_str(), W_OK) == 0) {
        break;
      }
      std::this_thread::sleep_for(PWM_EXPORT_RETRY_DELAY);
    }
    write_sysfs(duty_path, "0");
    write_sysfs(channel_path + "/period", std::to_string(PULSE_PERIOD_NS));
    write_sysfs(channel_path + "/enable", "1");
    duty_file = std::fopen(duty_path.c_str(), "w");
    if (duty_file == nullptr) {
      throw std::runtime_error("failed opening " + duty_path);
    }
    halt();
  }

  ~PulseChannel()
  {
    halt();
    std::fclose(duty_file);
    try {
      write_sysfs(channel_path + "/enable", "0");
    } catch (const std::runtime_error &) {
    }
  }

  void drive(double wheel_speed, double max_wheel_speed)
  {
    double fraction = std::clamp(wheel_speed / max_wheel_speed, -1.0, 1.0);
    if (reversed) {
      fraction = -fraction;
    }
    write_pulse_ns(
      PULSE_NEUTRAL_NS + static_cast<int64_t>(fraction * PULSE_HALF_RANGE_NS));
    command_fraction = wheel_speed >= 0.0 ? std::abs(fraction) : -std::abs(fraction);
    velocity = wheel_speed;
  }

  void halt()
  {
    write_pulse_ns(PULSE_NEUTRAL_NS);
    command_fraction = 0.0;
    velocity = 0.0;
  }

  double command_fraction = 0.0;
  double velocity = 0.0;
  double travel = 0.0;

private:
  void write_pulse_ns(int64_t pulse_ns)
  {
    if (pulse_ns == last_pulse_ns) {
      return;
    }
    std::rewind(duty_file);
    std::fprintf(duty_file, "%s", std::to_string(pulse_ns).c_str());
    std::fflush(duty_file);
    last_pulse_ns = pulse_ns;
  }

  const bool reversed;
  const std::string channel_path;
  std::FILE *duty_file;
  int64_t last_pulse_ns = 0;
};

class RcPulseMotorDriver : public rclcpp::Node
{
public:
  RcPulseMotorDriver()
  : Node("rc_pulse_motor_driver")
  {
    declare_parameter("max_wheel_speed", 1.0);
    for (const std::string side : {"left", "right"}) {
      declare_parameter(side + "_pwm_chip", 0);
      declare_parameter(side + "_pwm_channel", 0);
      declare_parameter(side + "_reversed", false);
    }

    max_wheel_speed = get_parameter("max_wheel_speed").as_double();
    for (size_t side_index = 0; side_index < wheels.size(); ++side_index) {
      const std::string side = side_index == 0 ? "left" : "right";
      wheels[side_index] = std::make_unique<PulseChannel>(
        get_parameter(side + "_pwm_chip").as_int(),
        get_parameter(side + "_pwm_channel").as_int(),
        get_parameter(side + "_reversed").as_bool());
    }

    last_command_time = now();
    drive_subscription = create_subscription<Drive>(
      "platform/motors/cmd_drive", rclcpp::SensorDataQoS(),
      [this](const Drive & message) {on_drive(message);});
    feedback_publisher =
      create_publisher<Feedback>("platform/motors/feedback", rclcpp::SensorDataQoS());
    feedback_timer = create_wall_timer(
      std::chrono::duration<double>(1.0 / FEEDBACK_RATE_HZ), [this]() {publish_feedback();});
  }

private:
  void on_drive(const Drive & message)
  {
    last_command_time = now();
    commanded_mode = message.mode;
    std::array<double, 2> wheel_speeds;
    if (message.mode == Drive::MODE_VELOCITY) {
      wheel_speeds = {message.drivers[0], message.drivers[1]};
    } else if (message.mode == Drive::MODE_PWM) {
      wheel_speeds = {
        message.drivers[0] * max_wheel_speed, message.drivers[1] * max_wheel_speed};
    } else {
      for (auto & wheel : wheels) {
        wheel->halt();
      }
      return;
    }
    for (size_t side_index = 0; side_index < wheels.size(); ++side_index) {
      wheels[side_index]->drive(wheel_speeds[side_index], max_wheel_speed);
    }
  }

  void publish_feedback()
  {
    const auto current_time = now();
    const bool has_timed_out =
      (current_time - last_command_time).seconds() > COMMAND_TIMEOUT_SECONDS;
    if (has_timed_out) {
      for (auto & wheel : wheels) {
        wheel->halt();
      }
    }

    Feedback message;
    message.header.stamp = current_time;
    message.commanded_mode = commanded_mode;
    message.actual_mode = has_timed_out ? Drive::MODE_NONE : commanded_mode;
    for (size_t side_index = 0; side_index < wheels.size(); ++side_index) {
      wheels[side_index]->travel += wheels[side_index]->velocity / FEEDBACK_RATE_HZ;
      message.drivers[side_index].duty_cycle = wheels[side_index]->command_fraction;
      message.drivers[side_index].measured_velocity = wheels[side_index]->velocity;
      message.drivers[side_index].measured_travel = wheels[side_index]->travel;
    }
    feedback_publisher->publish(message);
  }

  double max_wheel_speed;
  std::array<std::unique_ptr<PulseChannel>, 2> wheels;
  int8_t commanded_mode = Drive::MODE_NONE;
  rclcpp::Time last_command_time;
  rclcpp::Subscription<Drive>::SharedPtr drive_subscription;
  rclcpp::Publisher<Feedback>::SharedPtr feedback_publisher;
  rclcpp::TimerBase::SharedPtr feedback_timer;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<RcPulseMotorDriver>());
  rclcpp::shutdown();
  return 0;
}

File: apps/robot/src/bin/rc_pulse_motor_driver.rs
use std::{
    path::Path,
    time::{Duration, Instant},
};

use oxidros::prelude::*;
use tokio::time::interval;

use robot::{
    driver::command_is_stale,
    platform_msgs::robot_platform_msgs::msg::{Drive, Feedback},
};
use robot_control::parameters::{bool_parameter, f64_parameter, i64_parameter};
use robot_drivers::sysfs_pwm::PulseChannel;

const SYSFS_PWM_ROOT: &str = "/sys/class/pwm";
const FEEDBACK_RATE_HZ: f64 = 50.0;
const COMMAND_TIMEOUT_SECONDS: f64 = 0.5;
const DEFAULT_MAX_WHEEL_SPEED: f64 = 1.0;

type BoxError = Box<dyn std::error::Error + Send + Sync>;

struct Wheel {
    channel: PulseChannel,
    command_fraction: f64,
    velocity: f64,
    travel: f64,
}

impl Wheel {
    fn new(channel: PulseChannel) -> Self {
        Self {
            channel,
            command_fraction: 0.0,
            velocity: 0.0,
            travel: 0.0,
        }
    }

    fn drive(&mut self, wheel_speed: f64, max_wheel_speed: f64) -> std::io::Result<()> {
        self.command_fraction = self.channel.drive(wheel_speed / max_wheel_speed)?;
        self.velocity = wheel_speed;
        Ok(())
    }

    fn halt(&mut self) -> std::io::Result<()> {
        self.channel.halt()?;
        self.command_fraction = 0.0;
        self.velocity = 0.0;
        Ok(())
    }
}

fn feedback(wheels: &mut [Wheel; 2], commanded_mode: i8, has_timed_out: bool) -> Feedback {
    let (sec, nanosec) = oxidros::clock::Clock::new()
        .unwrap()
        .get_now()
        .map(|now| (now.as_secs() as i32, now.subsec_nanos()))
        .unwrap_or_default();
    let mut message = Feedback::default();
    message.header.stamp.sec = sec;
    message.header.stamp.nanosec = nanosec;
    message.commanded_mode = commanded_mode;
    message.actual_mode = if has_timed_out {
        Drive::MODE_NONE
    } else {
        commanded_mode
    };
    for (wheel, driver) in wheels.iter_mut().zip(message.drivers.iter_mut()) {
        wheel.travel += wheel.velocity / FEEDBACK_RATE_HZ;
        driver.duty_cycle = wheel.command_fraction as f32;
        driver.measured_velocity = wheel.velocity as f32;
        driver.measured_travel = wheel.travel as f32;
    }
    message
}

#[tokio::main]
async fn main() -> Result<(), BoxError> {
    init_ros_logging("rc_pulse_motor_driver");
    let context = Context::new()?;
    let node = context.create_node("rc_pulse_motor_driver", None)?;

    let (max_wheel_speed, channel_settings) = {
        let parameters = node.create_parameter_server()?;
        let store = parameters.params.read();
        (
            f64_parameter(&store, "max_wheel_speed", DEFAULT_MAX_WHEEL_SPEED),
            ["left", "right"].map(|side| {
                (
                    i64_parameter(&store, &format!("{side}_pwm_chip"), 0) as u32,
                    i64_parameter(&store, &format!("{side}_pwm_channel"), 0) as u32,
                    bool_parameter(&store, &format!("{side}_reversed"), false),
                )
            }),
        )
    };

    let sysfs_root = Path::new(SYSFS_PWM_ROOT);
    let mut wheels = [
        Wheel::new(PulseChannel::new(
            sysfs_root,
            channel_settings[0].0,
            channel_settings[0].1,
            channel_settings[0].2,
        )?),
        Wheel::new(PulseChannel::new(
            sysfs_root,
            channel_settings[1].0,
            channel_settings[1].1,
            channel_settings[1].2,
        )?),
    ];

    let mut cmd_drive =
        node.create_subscriber::<Drive>("platform/motors/cmd_drive", Some(Profile::sensor_data()))?;
    let feedback_publisher =
        node.create_publisher::<Feedback>("platform/motors/feedback", Some(Profile::sensor_data()))?;

    let mut commanded_mode = Drive::MODE_NONE;
    let mut last_command = Instant::now();
    let mut feedback_tick = interval(Duration::from_secs_f64(1.0 / FEEDBACK_RATE_HZ));
    tracing::info!("driving ESC pulses via {SYSFS_PWM_ROOT}");
    let ctrl_c = tokio::signal::ctrl_c();
    tokio::pin!(ctrl_c);
    loop {
        tokio::select! {
            _ = &mut ctrl_c => break,
            message = cmd_drive.recv() => {
                let command = message?.sample;
                last_command = Instant::now();
                commanded_mode = command.mode;
                match command.mode {
                    Drive::MODE_VELOCITY => {
                        for (wheel, wheel_speed) in wheels.iter_mut().zip(command.drivers) {
                            wheel.drive(wheel_speed as f64, max_wheel_speed)?;
                        }
                    }
                    Drive::MODE_PWM => {
                        for (wheel, fraction) in wheels.iter_mut().zip(command.drivers) {
                            wheel.drive(fraction as f64 * max_wheel_speed, max_wheel_speed)?;
                        }
                    }
                    _ => {
                        for wheel in &mut wheels {
                            wheel.halt()?;
                        }
                    }
                }
            }
            _ = feedback_tick.tick() => {
                let has_timed_out = command_is_stale(
                    last_command.elapsed().as_secs_f64(), COMMAND_TIMEOUT_SECONDS);
                if has_timed_out {
                    for wheel in &mut wheels {
                        wheel.halt()?;
                    }
                }
                feedback_publisher.send(&feedback(&mut wheels, commanded_mode, has_timed_out))?;
            }
        }
    }

    for wheel in &mut wheels {
        wheel.halt()?;
    }
    Ok(())
}
